src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs

```diff
@@ -122,6 +122,9 @@ public static unsafe ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
             if (RuntimeHelpers.IsReferenceOrContainsReferences<TTo>())
                 ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(TTo));
 
+            if (typeof(TFrom) == typeof(TTo))
```

Comment:
This change would be a regression for shared generic code. `TFrom` and `TTo` may require generic dictionary lookups.

The identical types should be covered by `if (fromSize == toSize)` path below. Is there anything preventing the JIT from generating the optimal code for that path?